draft-ietf-quic-qcram.md

```diff
@@ -127,23 +127,180 @@ considered blocked by the decoder and can not be processed until all entries in
 the range `[1, Depends Index]` have been added.  While blocked, header
 field data MUST remain in the blocked stream's flow control window.
 
-# HPACK extensions
+# Wire Format
+
+QCRAM instructions occur on three stream types, each of which uses a separate
+instruction space:
+
+ - Table updates are carried by HEADERS frames on the control stream, as defined
+   by {{QUIC-HTTP}}.  Frames on this stream modify the dynamic table state
+   without generating output to any particular request.
+ - Acknowledgement of header frame processing is carried by HEADER_ACK frames,
+   running from decoder to encoder.
+ - Finally, the contents of HEADERS and PUSH_PROMISE frames on request streams
+   reference the QPACK table state.
+
+This section describes the instructions which are possible on each stream type.
+
+In order to ensure table consistency and simplify update management, all table
+updates occur on the control stream rather than on request streams. Request
+streams contain only header blocks, which do not modify the state of the table.
+
+## Primitives
+
+The prefixed integer from Section 5.1 of [RFC7541] is used heavily throughout
+this document.  The string literal, defined by Section 5.2 of [RFC7541], is used
+with the following modification.
+
+HPACK defines string literals to begin on a byte boundary.  They begin with a
+single flag (indicating whether the string is Huffman-coded), followed by the
+Length encoded as a 7-bit prefix integer, and finally Length octets of data.
+
+QCRAM permits strings to begin other than on a byte boundary.  An "N-bit prefix
```

Comment:
Yes, I wrestled with that, too.  I'm inclined to say that once we start changing the instruction space, wire compatibility isn't a goal.  And that's the choice we need to make about the path forward.  You're correct that it's only a four-byte prefix, but that can represent header names up to 15 octets long; a surprising number of registered headers are fifteen octets or fewer (though in fairness, many of these are already in the static table or could be if we update it).

The other advantage of this is that, if we want to depart from HPACK one more step, it enables us to zero-base the tables rather than one-base the tables.

Reply:
My two cents. 

I think this and other PRs paint a pretty clear picture of moving further away from HPACK looks like.   There are some wins here for sure, but I'm not sure any of them are large enough to convince me that the moving away from HPACK is a good trade-off.   I *think* I heard the opposite at the Seattle interim, folks seemed to be favoring pursuit of the smallest changes with the biggest impact.   If the bar is that major changes should come with a big win, I'm not convinced that has been demonstrated.

I don't feel super strongly, but I'd feel a lot better with feedback from the broader WG.   Are we ok with making QPACK largely a re-write, or do we value the possibility to share implementation bits with HPACK?

Reply:
From my perspective, the encoding of instructions is not a big point of divergence.  As this is the top-level of the encoder, and we need new instructions anyway, I found that you need to have a different parser at this layer.  Given that the benefits are clear, I don't see why we shouldn't do it.